tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.stderr

```diff
@@ -0,0 +1,58 @@
+error[E0798]: arguments for `C-cmse-nonsecure-call` function too large to pass via registers
+  --> $DIR/params-via-stack.rs:24:5
+   |
+LL |     f1: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u32, u32) -> u32,
```

Comment:
we considered that but ultimately decided against that because 

- there is no risk of actual incorrect behavior (given the check at the call site)
- we don't know how to perform that check exhaustively

Is it feasible to disallow values of such types to ever exist? (e.g. in parameter lists, in types, as the output of `core::mem::transmute`).

We could consider adding lints later (e.g. an argument count of more than 4 is always incorrect).

Reply:
It could be done in `lower_fn_ty`. That's the one central place where function signatures are handled and we already have type information available, so you should be able to get the layout of the arguments and the return type.

Reply:
one thing I ran into earlier is that it's hard to unwrap `repr(transparent)` wrappers to get at the actual inner type. I was not able to do that earlier with just a `tcx` because it does not implement the required traits. In theory generics are also a problem I think though we could disallow them on these functions. Are there examples of where the `repr(transparant)` unwrapping  happens on HIR elsewhere? A point in favor of performing the check earlier would be that the error message shows up when running `cargo check` (currently it only shows when running `cargo build` or similar).

also in general, should we disallow a type that cannot be used (in any useful way) from being created (instead of just restricting its use). I can't really think of analogous cases either way.

Reply:
Yea we do try to reject types like this early and often just replace them with the error type after an error has been reported.

You don't need to process the HIR. `lower_fn_ty` converts the HIR to a `Ty<'tcx>` representation, so you can process it right after that happens.

Reply:
also to be clear: do you suggest to additionally error in `lower_fn_ty` if we can detect it, or to just try our best there, replace with some error type placeholder (no idea how to do that exactly but I'll have a look) and then still perform the check at codegen time if something slipped through the cracks? Or should all checking happen in `lower_fn_ty`?

Reply:
> that function just cannot be called in a non-monomorphic context.

you can just reject generic `cmse` ABI function signatures for now. The PR as is introduces new post-monomorphization errors that we usually try to avoid by adding appropriate trait bounds. Until you have a use for generic `csme` ABI functions, we should avoid them

Reply:
conveniently generics are not allowed on function pointers anyway, nor is `impl Trait`, but I'm struggling with cases like this 

```rust
pub fn test<T: Copy>(
    t: T,
    f: extern "C-cmse-nonsecure-call" fn(T, u32, u32, u32) -> u64,
) -> u64 {
    f(t, 2, 3, 4)
}
```

is it reliable to use `tcx.layout_of` returning `Err(_)` to indicate that a generic was used (and that hence the layout cannot be determined)? or is there some more idiomatic way to figure out whether there is any (nested) generic variable used in a type?

Reply:
Yea, relying on layout_of suffices